src/wayland/compositor/handlers.rs

```diff
@@ -118,24 +115,24 @@ where
 }
 
 fn implement_surface<R, Impl>(
-    surface: NewResource<wl_surface::WlSurface>,
+    surface: Main<wl_surface::WlSurface>,
     log: ::slog::Logger,
     implem: Rc<RefCell<Impl>>,
 ) -> wl_surface::WlSurface
 where
     R: Default + 'static,
     Impl: FnMut(SurfaceEvent, wl_surface::WlSurface, CompositorToken<R>) + 'static,
 {
-    let surface = surface.implement_closure(
-        {
-            let mut implem = SurfaceImplem::make(log, implem);
-            move |req, surface| implem.receive_surface_request(req, surface)
-        },
-        Some(|surface| SurfaceData::<R>::cleanup(&surface)),
-        SurfaceData::<R>::new(),
-    );
-    SurfaceData::<R>::init(&surface);
+    surface.quick_assign({
+        let mut implem = SurfaceImplem::make(log, implem);
+        move |surface, req, _| implem.receive_surface_request(req, surface.deref().clone())
+    });
+    surface.assign_destructor(Filter::new(|surface, _, _| SurfaceData::<R>::cleanup(&surface)));
     surface
+        .as_ref()
+        .user_data()
+        .set(|| SurfaceData::<R>::init(&surface));
+    surface.deref().clone()
```

Comment:
Is this translation about right?

Reply:
Hmm, no. The `user_data()` should be `set_threadsafe(|| SurfaceData::<R>::new())`, and then you should keep the `SurfaceData::<R>::init(&surface)`.

Reply:
Okay, I was thinking since there was previously `implement_closure` and `implement_threadsafe`, only the latter would result in `user_data().set_threadsafe()`. How do I determine which set fn to use?

Reply:
Also, this will require `R: Send`. Is that an expected change?

Reply:
I've pushed another commit assuming that `R: Send` is okay and all `.set`s need to be `.set_threadsafe`s.

Reply:
I'm not entirely sure it's good to require `R: Send`, but it's more in line with the internal implementation of `SurfaceData` which is done in terms of mutexes. Let's go with that for now.